client/client.go

```diff
@@ -5,42 +5,101 @@ import (
 	"fmt"
 	"io"
 	"net/http"
+	"time"
 
 	"github.com/deepmap/oapi-codegen/pkg/securityprovider"
-	"github.com/hashicorp/go-cleanhttp"
 	"github.com/hashicorp/go-retryablehttp"
 	"github.com/pkg/errors"
+	"golang.org/x/time/rate"
 )
 
+const (
+	rpm = 600
+
+	// last retry will wait 64 seconds
+	maxRetries   = 7
+	minRetryWait = 1 * time.Second
+	maxRetryWait = 65 * time.Second
+)
+
+type RateLimitedClient struct {
+	client      *http.Client
+	rateLimiter *rate.Limiter
+}
+
+func (c *RateLimitedClient) Do(req *http.Request) (*http.Response, error) {
+	if c.rateLimiter == nil {
+		return nil, errors.New("rate limiter not set")
+	}
+	ctx := req.Context()
+	err := c.rateLimiter.Wait(ctx)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
+func attentiveBackoff(min, max time.Duration, attemptNum int, resp *http.Response) time.Duration {
+	// Retry for rate limits and server errors.
+	if resp != nil && resp.StatusCode == http.StatusTooManyRequests {
+		// Check for a 'Retry-After' header.
+		retryAfter := resp.Header.Get("Retry-After")
+		if retryAfter != "" {
+			if retryAfterSeconds, err := time.ParseDuration(retryAfter + "s"); err == nil {
+				return retryAfterSeconds
+			}
+		}
+
+		// otherwise use the default backoff
+	}
+	return retryablehttp.DefaultBackoff(min, max, attemptNum, resp)
+}
+
 func New(ctx context.Context, apiKey, apiEndpoint, version string, opts ...ClientOption) (*ClientWithResponses, error) {
 	bearerTokenProvider, bearerTokenProviderErr := securityprovider.NewSecurityProviderBearerToken(apiKey)
 	if bearerTokenProviderErr != nil {
 		return nil, bearerTokenProviderErr
 	}
 
 	retryClient := retryablehttp.NewClient()
-	retryClient.RetryMax = 3
+
+	retryClient.RetryMax = maxRetries
+	retryClient.RetryWaitMin = minRetryWait
+	retryClient.RetryWaitMax = maxRetryWait
+	retryClient.Backoff = attentiveBackoff
 
 	base := retryClient.StandardClient()
 
 	// The generated client won't turn validation errors into actual errors, so we do this
-	// inside of a generic middleware.
-	base.Transport = Wrap(cleanhttp.DefaultTransport(), func(req *http.Request, next http.RoundTripper) (*http.Response, error) {
+	// // inside of a generic middleware.
+	base.Transport = Wrap(base.Transport, func(req *http.Request, next http.RoundTripper) (*http.Response, error) {
 		resp, err := next.RoundTrip(req)
+		if err != nil {
+			return resp, err
```

Comment:
an odd go style thing, if `err != nil`, we almost always return `nil, err`. 
`resp` should already be `nil` because of this, but it's just a consistency thing

edit- ah yes same as lisa's